include/deal.II/lac/sparse_matrix.templates.h

```diff
@@ -1328,6 +1352,8 @@ SparseMatrix<number>::precondition_Jacobi (Vector<somenumber>       &dst,
   AssertDimension (dst.size(), n());
   AssertDimension (src.size(), n());
 
+  check_nonzero_diagonal(*this);
```

Comment:
Fair point. I'll #ifdef out the body of the function so we don't have to do it everywhere where we call it. I'm also going to rename it `AssertNoZerosOnDiagonal` to make it more look like an assertion.